src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    /// <summary>
+    /// Provides a compile-time view of the current workspace solution.
+    /// Workaround for Razor projects which generate both design-time and compile-time source files.
+    /// TODO: remove https://github.com/dotnet/roslyn/issues/51678
+    /// </summary>
+    internal sealed class CompileTimeSolutionProvider : ICompileTimeSolutionProvider
+    {
+        [ExportWorkspaceServiceFactory(typeof(ICompileTimeSolutionProvider), WorkspaceKind.Host), Shared]
+        private sealed class Factory : IWorkspaceServiceFactory
+        {
+            [ImportingConstructor]
+            [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+            public Factory()
+            {
+            }
+
+            [Obsolete(MefConstruction.FactoryMethodMessage, error: true)]
+            public IWorkspaceService? CreateService(HostWorkspaceServices workspaceServices)
+                => new CompileTimeSolutionProvider(workspaceServices.Workspace);
+        }
+
+        private const string RazorEncConfigFileName = "RazorSourceGenerator.razorencconfig";
+
+        private readonly Workspace _workspace;
+        private readonly object _guard = new();
+
+        private Solution? _lazyCompileTimeSolution;
```

Comment:
Should this subscribe to WorkspaceChanged and null this field out? Otherwise this is potentially holding a reference alive to a Solution that won't be returned anymore?